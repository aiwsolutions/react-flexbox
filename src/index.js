/**
 * Copyright by AIWSolutions.
 */
/* eslint object-curly-spacing:0 no-underscore-dangle: 0, global-require: 0 */
if (!global._babelPolyfill) {
    require('babel-polyfill');
}
export Flex from './Flex';
export Box from './Box';
