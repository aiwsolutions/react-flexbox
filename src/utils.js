/**
 * Copyright by AIWSolutions.
 */
import _ from 'lodash';

export const MARGIN_PROPS = ['m', 'mt', 'mb', 'ml', 'mr'];
export const PADDING_PROPS = ['p', 'pt', 'pb', 'pl', 'pr'];

const GLOBAL_ATTRIBUTES = [
    'accessKey',
    'className',
    'contentEditable',
    'dir',
    'draggable',
    'hidden',
    'id',
    'lang',
    'spellCheck',
    'tabIndex',
    'title',
    'translate',
];

const EVENT_HANDLERS_REGEX = /^on[A-Z][a-z]*/;

const DATA_REGEX = /^data-[a-zA-Z]*/;

const ARIA_REGEX = /^aria-[a-zA-Z]*/;

function scale(value) {
    if (_.isNumber(value)) {
        return value * 8;
    }
    return value;
}

export function computeStyle(props) {
    const {
        mt, mr, mb, ml, m, pt, pr, pb, pl, p, data, ...others
    } = props;
    return {
        style: _.pickBy({
            marginTop: scale(mt),
            marginRight: scale(mr),
            marginBottom: scale(mb),
            marginLeft: scale(ml),
            margin: scale(m),
            paddingTop: scale(pt),
            paddingRight: scale(pr),
            paddingBottom: scale(pb),
            paddingLeft: scale(pl),
            padding: scale(p),
        }, value => !_.isUndefined(value)),
        others: _.pickBy(others, (value, key) =>
            _.includes(GLOBAL_ATTRIBUTES, key)
            || EVENT_HANDLERS_REGEX.test(key)
            || DATA_REGEX.test(key)
            || ARIA_REGEX.test(key)),
    };
}
