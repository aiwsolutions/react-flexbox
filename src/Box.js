import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { computeStyle, MARGIN_PROPS, PADDING_PROPS } from './utils';

/**
 * Represent a simple container which is not a Flex container.
 * This container has built in shouldComponentUpdate to prevent unwanted updates when there is no changes on
 * its property
 * There is also basic property filtering for invalid propery when passing property from its parent like this:
 *      <Box
 *          {...props}
 *      />
 *
 * @property {string} align     - Shorthand for css align-self, this allows the default alignment (or the one
 *                                specified by align-items) to be overridden for individual flex items.
 * @property {bool} auto        - Shorthand for css flex: 1 1 auto
 * @property {string | number} basis   - Shorthand for css flex-basis, it defines the default size of an element
 *                                before the remaining space is distributed
 * @property {number} col       - How many x of 1/12 of width of the parent container that this element should take.
 *                                This equals to css width: (col * 100 / 12)%
 * @property {number} grow      - Shorthand for css flex-grow, it defines what amount of the available
 *                                space inside the flex container the item should take up
 * @property {number} order     - Controls the order in which this Box appears in the Flex container
 * @property {number} shrink    - Shorthand for css flex-shrink, it defines the ability for a flex item
 *                                to shrink if necessary.
 * @property {object} style     - Element style
 */
class Box extends React.Component {
    static propTypes = {
        order: PropTypes.number,
        auto: PropTypes.bool,
        grow: PropTypes.number,
        shrink: PropTypes.number,
        basis: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        align: PropTypes.oneOf([
            'auto',
            'flex-start',
            'flex-end',
            'center',
            'baseline',
            'stretch',
        ]),
        col: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
        style: PropTypes.object,
        className: PropTypes.string,
    };

    // Should not get from propTypes as it will be trimmed
    static ownProps = [
        'order', 'auto', 'grow', 'shrink', 'basic', 'align', 'col', 'style', 'className',
        ...MARGIN_PROPS, ...PADDING_PROPS
    ];

    shouldComponentUpdate(nextProps) {
        return this.props.children !== nextProps.children
            || !_.isEqual(_.pick(this.props, Box.ownProps), _.pick(nextProps, Box.ownProps));
    }

    render() {
        const {
            children, order, auto, grow, shrink, basis, align, col, style, ...others
        } = this.props;
        const boxStyle = {
            WebkitOrder: order,
            order,
            WebkitBoxSizing: 'border-box',
            boxSizing: 'border-box',
            WebkitFlexGrow: grow,
            flexGrow: grow,
            WebkitFlexShrink: shrink,
            flexShrink: shrink,
            WebkitFlexBasis: basis,
            flexBasis: basis,
            WebkitAlignSelf: align,
            alignSelf: align,
        };

        if (auto) {
            boxStyle.WebkitFlex = '1 1 auto';
            boxStyle.flex = '1 1 auto';
        }

        if (col) {
            const widthPercentage = col * 100 / 12;
            boxStyle.width = `${widthPercentage.toFixed(2)}%`;
        }

        const computedStyle = computeStyle(others);
        return (
            <div
                style={Object.assign({}, boxStyle, computedStyle.style, style)}
                {...computedStyle.others}
            >
                {children}
            </div>
        );
    }
}

export default Box;
