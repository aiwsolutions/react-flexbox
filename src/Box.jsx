/**
 * Copyright by AIWSolutions.
 */
import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { computeStyle } from './utils';

class Box extends React.Component {
    static propTypes = {
        order: PropTypes.number,
        auto: PropTypes.bool,
        grow: PropTypes.number,
        shrink: PropTypes.number,
        basis: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        align: PropTypes.oneOf([
            'auto',
            'flex-start',
            'flex-end',
            'center',
            'baseline',
            'stretch',
        ]),
        col: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
        style: PropTypes.object,
    };

    shouldComponentUpdate(nextProps) {
        return this.props.children !== nextProps.children
            || !_.isEqual(_.pick(this.props, this.ownProps), _.pick(nextProps, this.ownProps));
    }

    ownProps = _.keys(Box.propTypes);

    render() {
        const {
            children, order, auto, grow, shrink, basis, align, col, style, ...others
        } = this.props;
        const boxStyle = {
            WebkitOrder: order,
            order,
            WebkitBoxSizing: 'border-box',
            boxSizing: 'border-box',
            WebkitFlexGrow: grow,
            flexGrow: grow,
            WebkitFlexShrink: shrink,
            flexShrink: shrink,
            WebkitFlexBasis: basis,
            flexBasis: basis,
            WebkitAlignSelf: align,
            alignSelf: align,
        };

        if (auto) {
            boxStyle.WebkitFlex = '1 1 auto';
            boxStyle.flex = '1 1 auto';
        }

        if (col) {
            const widthPercentage = (col * 100) / 12;
            boxStyle.width = `${widthPercentage.toFixed(2)}%`;
        }

        const computedStyle = computeStyle(others);
        return (
            <div
              style={Object.assign({}, boxStyle, computedStyle.style, style)}
              {...computedStyle.others}
            >
                {children}
            </div>
        );
    }
}

export default Box;
