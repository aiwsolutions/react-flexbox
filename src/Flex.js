import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { computeStyle, MARGIN_PROPS, PADDING_PROPS } from './utils';
import './flexbox.css';

/**
 * Represent a Flex container.
 * This container has built in shouldComponentUpdate to prevent unwanted updates when there is no changes on
 * its property
 * There is also basic property filtering for invalid propery when passing property from its parent like this:
 *      <Flex
 *          {...props}
 *      />
 *
 * @property {string} align     - Shorthand for css align-items, This defines the default behaviour for how flex 
 *                                items are laid out along the cross axis on the current line. 
 * @property {bool} auto        - Shorthand for css flex: 1 1 auto
 * @property {string | number} basis   - Shorthand for css flex-basis, it defines the default size of an element
 *                                before the remaining space is distributed
 * @property {string} className - By default this element will have className of 'flex' or 'inline-flex', this
 *                                property is for extra classes
 * @property {number} col       - How many x of 1/12 of width of the parent container that this element should take.
 *                                This equals to css width: (col * 100 / 12)%
 * @property {bool} fullWidth   - Shorthand for css width: 100%
 * @property {number} grow      - Shorthand for css flex-grow, it defines what amount of the available
 *                                space inside the flex container the item should take up
 * @property {bool} inline      - Make this Flex to be inline-flex which displays itself as inline
 * @property {string} justify   - Shorthand for css justify-content, it defines the alignment along the main axis.
 * @property {number} order     - Controls the order in which this Box appears in the Flex container
 * @property {bool} reverse     - Reverse the direction of aligning items
 * @property {number} shrink    - Shorthand for css flex-shrink, it defines the ability for a flex item
 *                                to shrink if necessary.
 * @property {object} style     - Element style
 * @property {string} wrap      - Shorthand for css flex-wrap, it allows the items to wrap as needed
 */
class Flex extends React.Component {
    static propTypes = {
        auto: PropTypes.bool,
        column: PropTypes.bool,
        reverse: PropTypes.bool,
        order: PropTypes.number,
        grow: PropTypes.number,
        shrink: PropTypes.number,
        wrap: PropTypes.oneOf([
            'nowrap',
            'wrap',
            'wrap-reverse',
        ]),
        justify: PropTypes.oneOf([
            'flex-start',
            'flex-end',
            'center',
            'space-between',
            'space-around',
        ]),
        align: PropTypes.oneOf([
            'flex-start',
            'flex-end',
            'center',
            'stretch',
            'baseline',
        ]),
        fullWidth: PropTypes.bool,
        inline: PropTypes.bool,
        col: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
        style: PropTypes.object,
        className: PropTypes.string,
    };

    static ownProps = [
        'auto', 'column', 'reverse', 'order', 'grow', 'shrink', 'wrap', 'justify', 'align', 'fullWidth',
        'inline', 'col', 'style', 'className', ...MARGIN_PROPS, ...PADDING_PROPS
    ];

    shouldComponentUpdate(nextProps) {
        return this.props.children !== nextProps.children
            || !_.isEqual(_.pick(this.props, Flex.ownProps), _.pick(nextProps, Flex.ownProps));
    }

    render() {
        const {
            children,
            auto,
            order,
            grow,
            shrink,
            basis,
            column,
            reverse,
            wrap,
            justify,
            align,
            fullWidth,
            style,
            col,
            inline,
            className,
            shouldUpdate,
            ...others
        } = this.props;

        /* eslint no-nested-ternary: 0 */
        const flexDirection = column ? (reverse ? 'column-reverse' : 'column')
            : (reverse ? 'row-reverse' : 'row');

        const flexStyle = {
            WebkitOrder: order,
            order,
            WebkitBoxSizing: 'border-box',
            boxSizing: 'border-box',
            WebkitFlexGrow: grow,
            flexGrow: grow,
            WebkitFlexShrink: shrink,
            flexShrink: shrink,
            WebkitFlexBasis: basis,
            flexBasis: basis,
            WebkitFlexDirection: flexDirection,
            flexDirection,
            WebkitFlexWrap: wrap,
            flexWrap: wrap,
            WebkitJustifyContent: justify,
            justifyContent: justify,
            WebkitAlignItems: align,
            alignItems: align,
        };

        if (auto) {
            flexStyle.WebkitFlex = '1 1 auto';
            flexStyle.flex = '1 1 auto';
        }

        if (fullWidth) {
            flexStyle.width = '100%';
        }
        if (col) {
            const widthPercentage = col * 100 / 12;
            flexStyle.width = `${widthPercentage.toFixed(2)}%`;
        }

        const computedStyle = computeStyle(others);
        const classString = inline ? 'inline-flex' : 'flex';
        return (
            <div
                className={classString + (className ? ` ${className}` : '')}
                style={Object.assign({}, flexStyle, computedStyle.style, style)}
                {...computedStyle.others}
            >
                {children}
            </div>
        );
    }
}

export default Flex;
