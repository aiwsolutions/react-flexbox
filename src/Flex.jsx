/**
 * Copyright by AIWSolutions.
 */
import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { computeStyle } from './utils';
import './flexbox.css';

class Flex extends React.Component {
    static propTypes = {
        column: PropTypes.bool,
        reverse: PropTypes.bool,
        wrap: PropTypes.oneOf([
            'nowrap',
            'wrap',
            'wrap-reverse',
        ]),
        justify: PropTypes.oneOf([
            'flex-start',
            'flex-end',
            'center',
            'space-between',
            'space-around',
        ]),
        align: PropTypes.oneOf([
            'flex-start',
            'flex-end',
            'center',
            'stretch',
            'baseline',
        ]),
        fullWidth: PropTypes.bool,
        inline: PropTypes.bool,
        col: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
        auto: PropTypes.bool,
        style: PropTypes.object,
    };


    shouldComponentUpdate(nextProps) {
        return this.props.children !== nextProps.children
            || !_.isEqual(_.pick(this.props, this.ownProps), _.pick(nextProps, this.ownProps));
    }

    ownProps = _.keys(Flex.propType);

    render() {
        const {
            children,
            auto,
            column,
            reverse,
            wrap,
            justify,
            align,
            fullWidth,
            style,
            col,
            inline,
            ...others
        } = this.props;

        /* eslint no-nested-ternary: 0 */
        const flexDirection = column ? (reverse ? 'column-reverse' : 'column')
            : (reverse ? 'row-reverse' : 'row');

        const flexStyle = {
            WebkitBoxSizing: 'border-box',
            boxSizing: 'border-box',
            WebkitFlexDirection: flexDirection,
            flexDirection,
            WebkitFlexWrap: wrap,
            flexWrap: wrap,
            WebkitJustifyContent: justify,
            justifyContent: justify,
            WebkitAlignItems: align,
            alignItems: align,
        };

        if (auto) {
            flexStyle.WebkitFlex = '1 1 auto';
            flexStyle.flex = '1 1 auto';
        }

        if (fullWidth) {
            flexStyle.width = '100%';
        }
        if (col) {
            const widthPercentage = (col * 100) / 12;
            flexStyle.width = `${widthPercentage.toFixed(2)}%`;
        }

        const computedStyle = computeStyle(others);
        return (
            <div
              className={inline ? 'inline-flex' : 'flex'}
              style={Object.assign({}, flexStyle, computedStyle.style, style)}
              {...computedStyle.others}
            >
                {children}
            </div>
        );
    }
}

export default Flex;
