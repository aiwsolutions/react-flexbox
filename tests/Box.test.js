import React from 'react';
import renderer from 'react-test-renderer';
import {Box} from '../src/index';

test('<Box> - default', () => {
    const comp = renderer.create(
        <Box/>
    );
    expect(comp.toJSON()).toMatchSnapshot();
})
