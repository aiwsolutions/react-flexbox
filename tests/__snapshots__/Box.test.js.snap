// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box> - default 1`] = `
<div
  style={
    Object {
      "WebkitAlignSelf": undefined,
      "WebkitBoxSizing": "border-box",
      "WebkitFlexBasis": undefined,
      "WebkitFlexGrow": undefined,
      "WebkitFlexShrink": undefined,
      "WebkitOrder": undefined,
      "alignSelf": undefined,
      "boxSizing": "border-box",
      "flexBasis": undefined,
      "flexGrow": undefined,
      "flexShrink": undefined,
      "order": undefined,
    }
  }
/>
`;
